Source/Core/VideoBackends/Vulkan/ObjectCache.cpp

```diff
@@ -135,33 +135,74 @@ GetVulkanDepthStencilState(const DepthStencilState& state)
   };
 }
 
-static VkPipelineColorBlendAttachmentState GetVulkanAttachmentBlendState(const BlendState& state)
+static VkPipelineColorBlendAttachmentState GetVulkanAttachmentBlendState(const BlendingState& state)
 {
-  VkPipelineColorBlendAttachmentState vk_state = {
-      state.blend_enable,     // VkBool32                                  blendEnable
-      state.src_blend,        // VkBlendFactor                             srcColorBlendFactor
-      state.dst_blend,        // VkBlendFactor                             dstColorBlendFactor
-      state.blend_op,         // VkBlendOp                                 colorBlendOp
-      state.src_alpha_blend,  // VkBlendFactor                             srcAlphaBlendFactor
-      state.dst_alpha_blend,  // VkBlendFactor                             dstAlphaBlendFactor
-      state.alpha_blend_op,   // VkBlendOp                                 alphaBlendOp
-      state.write_mask        // VkColorComponentFlags                     colorWriteMask
-  };
+  static constexpr std::array<VkBlendFactor, 8> src_factors = {
+      {VK_BLEND_FACTOR_ZERO, VK_BLEND_FACTOR_ONE, VK_BLEND_FACTOR_DST_COLOR,
+       VK_BLEND_FACTOR_ONE_MINUS_DST_COLOR, VK_BLEND_FACTOR_SRC_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_SRC_ALPHA, VK_BLEND_FACTOR_DST_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_DST_ALPHA}};
+  static constexpr std::array<VkBlendFactor, 8> src_factors_dualsrc = {
+      {VK_BLEND_FACTOR_ZERO, VK_BLEND_FACTOR_ONE, VK_BLEND_FACTOR_DST_COLOR,
+       VK_BLEND_FACTOR_ONE_MINUS_DST_COLOR, VK_BLEND_FACTOR_SRC1_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_SRC1_ALPHA, VK_BLEND_FACTOR_DST_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_DST_ALPHA}};
+  static constexpr std::array<VkBlendFactor, 8> dst_factors = {
+      {VK_BLEND_FACTOR_ZERO, VK_BLEND_FACTOR_ONE, VK_BLEND_FACTOR_SRC_COLOR,
+       VK_BLEND_FACTOR_ONE_MINUS_SRC_COLOR, VK_BLEND_FACTOR_SRC_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_SRC_ALPHA, VK_BLEND_FACTOR_DST_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_DST_ALPHA}};
+  static constexpr std::array<VkBlendFactor, 8> dst_factors_dualsrc = {
+      {VK_BLEND_FACTOR_ZERO, VK_BLEND_FACTOR_ONE, VK_BLEND_FACTOR_SRC_COLOR,
+       VK_BLEND_FACTOR_ONE_MINUS_SRC_COLOR, VK_BLEND_FACTOR_SRC1_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_SRC1_ALPHA, VK_BLEND_FACTOR_DST_ALPHA,
+       VK_BLEND_FACTOR_ONE_MINUS_DST_ALPHA}};
+
+  VkPipelineColorBlendAttachmentState vk_state = {};
+  vk_state.blendEnable = static_cast<VkBool32>(state.blendenable);
+  vk_state.colorBlendOp = state.subtract ? VK_BLEND_OP_REVERSE_SUBTRACT : VK_BLEND_OP_ADD;
+  vk_state.alphaBlendOp = state.subtractAlpha ? VK_BLEND_OP_REVERSE_SUBTRACT : VK_BLEND_OP_ADD;
+  vk_state.srcColorBlendFactor =
```

Comment:
It might be a little nicer if all these things depended on 'use dual src' was in a single if/else block